### PR TITLE
ElementReference FocusAsync API

### PR DESCRIPTION
**Pull request by MackinnonBuck**
Summary of the changes:

- [X] Added `FocusAsync()` extension method to `ElementReference` that focuses the element associated with the reference.
- [X] Added tests to verify that `FocusAsync()` works properly.
- [x] Improved API to omit the `IJSRuntime` argument from `FocusAsync()`

Addresses #22892

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=23316) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=23316) before we can accept your contribution.<br/><br/>:x: MackinnonBuck [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=23316)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=23316) it.</sub>

**Comment by SteveSandersonMS** [comment]
This looks great, and it's awesome to see the E2E test in there as well!

I'm guessing this is marked draft because you're going on to do the "omit JSRuntime arg" bit, so I'll wait for updates :)

**Review by javiercn — COMMENTED** [review]
Looks great!

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj [hunk 1]
```diff
@@ -62,4 +57,8 @@
     <NuspecProperty Include="PackageThirdPartyNoticesFile=$(PackageThirdPartyNoticesFile)" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\JSInterop\Microsoft.JSInterop\src\Microsoft.JSInterop.csproj" />
```
Comment (pranavkm) [inline src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Doesn't this project already reference JSInterop? You shouldn't need to do add this reference

Comment (SteveSandersonMS) [inline src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Ah, unfortunately I didn't realise this would be needed before.

We want to avoid referencing JS interop concepts directly from the `.Components` package because that's meant to be independent of web technology. There are other forms of component rendering that don't involve web browsers or the concept of JavaScript at all.

I know it's quirky that `ElementReference` is a concept inside `.Components`, but the intention is that although for HTML rendering an "element" maps to an HTML element, for other forms of rendering an "element" maps to whatever is the primitive unit of UI in that technology. For example, in our Xamarin Forms renderer, the "element" concept corresponds to an underlying native Xamarin Forms control, and in this form of rendering, there is no JavaScript and hence no JS interop.

I'm not sure if this undermines the whole idea of eliminating the `IJSRuntime` parameter to `focus`. Maybe there might still be some way of doing it, for example by storing the `IServiceProvider` on `ElementReference` so that the `Focus` extension method (which is in the `.Web` project, which is allowed to have a concept of JS interop) can dynamically look up the `IJSRuntime` from the service provider. For the avoidance of doubt, we don't want to expose the `Renderer` instance to application code.

Maybe there's another cleverer solution still, not sure!

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 2]
```diff
@@ -369,6 +371,16 @@ internal void TrackReplacedEventHandlerId(ulong oldEventHandlerId, ulong newEven
...
         }
 
+        /// <summary>
+        /// Gets the <see cref="IJSRuntime"/> service from the <see cref="Renderer"/>'s service provider.
+        /// </summary>
+        /// <remarks>
+        /// The returned <see cref="IJSRuntime"/> may be <c>null</c> if the <see cref="Renderer"/>'s service
+        /// provider does not have one.
+        /// </remarks>
+        internal IJSRuntime GetJSRuntime()
```
Comment (pranavkm) [inline src/Components/Components/src/RenderTree/Renderer.cs:381]:
```suggestion
        internal IJSRuntime JSRuntime { get; } = _serviceProvider.GetRequiredService<IJSRuntime>();
```

if this doesn't work, you could assign the property in the constructor.

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 3]
```diff
@@ -369,6 +371,16 @@ internal void TrackReplacedEventHandlerId(ulong oldEventHandlerId, ulong newEven
             _eventHandlerIdReplacements.Add(oldEventHandlerId, newEventHandlerId);
         }
 
+        /// <summary>
+        /// Gets the <see cref="IJSRuntime"/> service from the <see cref="Renderer"/>'s service provider.
+        /// </summary>
+        /// <remarks>
+        /// The returned <see cref="IJSRuntime"/> may be <c>null</c> if the <see cref="Renderer"/>'s service
```
Comment (pranavkm) [inline src/Components/Components/src/RenderTree/Renderer.cs:378]: This shouldn't happen in actual apps.

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 4]
```diff
@@ -369,6 +371,16 @@ internal void TrackReplacedEventHandlerId(ulong oldEventHandlerId, ulong newEven
             _eventHandlerIdReplacements.Add(oldEventHandlerId, newEventHandlerId);
         }
 
+        /// <summary>
+        /// Gets the <see cref="IJSRuntime"/> service from the <see cref="Renderer"/>'s service provider.
```
Comment (pranavkm) [inline src/Components/Components/src/RenderTree/Renderer.cs:375]: We generally do not document internal APIs. But if you would like to, I'd leave it at 
```suggestion
        /// Gets the <see cref="IJSRuntime"/> instance.
```

The implementation details aren't useful to consumers and might change which would leave an out-of-date doc

File: src/Components/Server/test/ElementReferenceJsonConverterTest.cs [hunk 5]
```diff
@@ -4,19 +4,30 @@
...
     public class ElementReferenceJsonConverterTest
     {
-        private readonly ElementReferenceJsonConverter Converter = new ElementReferenceJsonConverter();
+        private readonly IJSRuntime TestRuntime;
+
+        private readonly ElementReferenceJsonConverter Converter;
+
+        public ElementReferenceJsonConverterTest()
+        {
+            TestRuntime = new TestJsRuntime();
```
Comment (pranavkm) [inline src/Components/Server/test/ElementReferenceJsonConverterTest.cs:22]: This is fine as-is, but there's a cheaper way to do this especially if you aren't doing anything with the passed in value:

```suggestion
            TestRuntime = Mock.Of<IJSRuntime>();
```

File: src/Components/Web.JS/src/DomWrapper.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import '@microsoft/dotnet-js-interop';
+
+export const internalFunctions = {
```
Comment (pranavkm) [inline src/Components/Web.JS/src/DomWrapper.ts:3]: `domFunctions`?

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Nope, see comment above.

**Review by pranavkm — COMMENTED** [review]
This is really good! Just a few more things to address and we can get this in.

File: src/Components/Components/src/ElementReference.cs [hunk 7]
```diff
@@ -23,13 +24,20 @@ namespace Microsoft.AspNetCore.Components
...
         public string Id { get; }
 
-        public ElementReference(string id)
+        /// <summary>
+        /// Gets the <see cref="IJSRuntime"/> instance used to perform JS interop calls associated with
+        /// this <see cref="ElementReference"/>.
+        /// </summary>
+        public IJSRuntime JSRuntime { get; }
+
+        public ElementReference(string id, IJSRuntime jsRuntime)
```
Comment (pranavkm) [inline src/Components/Components/src/ElementReference.cs:33]: Since this involves adding new API, you will have to regenerate the reference assembly: https://github.com/dotnet/aspnetcore/blob/master/docs/ReferenceAssemblies.md#when-changing-public-api

Comment (pranavkm) [reply src/Components/Components/src/ElementReference.cs:33]: Could you also add a doc comment for the constructor?

File: src/Components/Components/src/ElementReference.cs [hunk 8]
```diff
@@ -23,13 +24,20 @@ namespace Microsoft.AspNetCore.Components
         /// </remarks>
         public string Id { get; }
 
-        public ElementReference(string id)
+        /// <summary>
+        /// Gets the <see cref="IJSRuntime"/> instance used to perform JS interop calls associated with
```
Comment (pranavkm) [inline src/Components/Components/src/ElementReference.cs:28]:
```suggestion
        /// Gets the <see cref="IJSRuntime"/> instance.
```

should suffice.

File: src/Components/Web/src/ElementReferenceExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+{
+    public static class ElementReferenceExtensions
+    {
+        /// <summary>
+        /// Gives focus to an element given its <see cref="ElementReference"/>.
+        /// </summary>
+        /// <param name="elementReference">A reference to the element to focus.</param>
+        /// <returns>The <see cref="ValueTask"/> representing the asynchronous focus operation.</returns>
+        public static ValueTask FocusAsync(this ElementReference elementReference)
+            => elementReference.JSRuntime.InvokeVoidAsync(DomWrapperInterop.Focus, elementReference);
```
Comment (pranavkm) [inline src/Components/Web/src/ElementReferenceExtensions.cs:17]: This is being extra cautious, but could you add a null check for `ElementReference.JSRuntime`? Just to guard against

```
await default(ElementReference).FocusAsync()
```

File: src/Components/Components/src/ElementReference.cs [hunk 10]
```diff
@@ -23,13 +24,20 @@ namespace Microsoft.AspNetCore.Components
...
+        /// this <see cref="ElementReference"/>.
+        /// </summary>
+        public IJSRuntime JSRuntime { get; }
+
+        public ElementReference(string id, IJSRuntime jsRuntime)
         {
             Id = id;
+            JSRuntime = jsRuntime;
```
Comment (pranavkm) [inline src/Components/Components/src/ElementReference.cs:36]:
```suggestion
            JSRuntime = jsRuntime ?? throw new ArgumentNullException(nameof(jsRuntime));
```

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj (hunk 1 above)
Comment (MackinnonBuck) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: It appears not. @SteveSandersonMS mentioned above that it would be best not to reference JS interop from the `Components` package, so I'll have to look into another solution. I can try changing `ElementReference` from having an `IJSInterop` property to having an `IServiceProvider` property instead, although that might get tricky with how `ElementReference`s are currently serialized/deserialized using `ElementReferenceJsonConverter`.

Comment (pranavkm) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: image

Comment (pranavkm) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: I'm not really sure what's it there for though.

File: src/Components/Web/src/ElementReferenceExtensions.cs (hunk 9 above)
Comment (MackinnonBuck) [reply src/Components/Web/src/ElementReferenceExtensions.cs:17]: Do you think it would be good to throw an exception if the null check fails, instead of silently not focusing? Or maybe log something?

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: I didn’t realise the reference was there. I guess it would be difficult to remove now. 

But in any case we shouldn’t assume in this project that the service provider will have any IJSRuntime. It won’t in the case of Mobile Blazor Bindings.

Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Would be interesting to understand how the reference got there. It might be an unintended consequence of the SDK setup or something. It’s too late here for me to dig into it right now, but I wonder if this project really makes use of that reference at all.

Comment (pranavkm) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Building on @MackinnonBuck's suggestion, we could avoid exposing this as part of Component's API surface. It might make it easier to break this if we'd like to before 5.0:

```
internal object? JSRuntime { get; set; }

Components.Web
{
  public static IJSRuntime GetJSRuntime(this ElementReference reference) => reference.JSRuntime ?? throw new PlatformNotSupportedException(..);
}

File: src/Components/Web/src/ElementReferenceExtensions.cs (hunk 9 above)
Comment (pranavkm) [reply src/Components/Web/src/ElementReferenceExtensions.cs:17]: Throwing an exception sounds great.

File: src/Components/Components/src/RenderTree/Renderer.cs (hunk 2 above)
Comment (pranavkm) [reply src/Components/Components/src/RenderTree/Renderer.cs:381]: Based on Steve's comment, I guess this should continue to use `GetService` in the event there are environments this isn't supported.

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: If we’re happy to add an InternalsVisibleTo then yes, I agree that avoids any impact on public API surface.

Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: But actually I’m not sure how the renderer is meant to supply that without knowing about the IJSRuntime.

Comment (pranavkm) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Let's do that. We could consider making a breaking change for 5.0 to move the package reference to Components.Web. I think all we really need is the type of IJSRuntime to lookup from DI which does not require the reference.

Comment (pranavkm) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: Alternatively, let's get this change in and we can figure it out in a follow up. The package reference exists, and we're only making it slightly worse 👍

Comment (MackinnonBuck) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: I've been looking into the suggestion of storing the `IServiceProvider` inside the `ElementReference` and I don't think this can easily be done since each `ElementReferenceJsonConverter` would need to be initialized with an `IServiceProvider` in order to deserialize `ElementReference`s. And it isn't always possible to provide one (e.g. in the case of `DefaultWebAsssemblyJSRuntime`).

Comment (MackinnonBuck) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: So just to clarify, are we still adding `ElementReference.JSRuntime` to the public API, are are we hiding it?

Comment (SteveSandersonMS) [reply src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:61]: I just checked this with Pranav and he and I think the `IServiceProvider` option is likely the cleanest one. @pranavkm can you provide pointers on how to get access to the service provider from inside `DefaultWebAssemblyJSRuntime`?

File: src/Components/Components/src/Microsoft.AspNetCore.Components.csproj [hunk 11]
```diff
@@ -33,18 +33,13 @@
   </ItemGroup>
 
   <Target Name="_GetNuspecDependencyPackageVersions">
-    <MSBuild Targets="_GetPackageVersionInfo"
```
Comment (pranavkm) [inline src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:36]: Could you undo changes to this file?

File: src/Components/Server/test/ElementReferenceJsonConverterTest.cs [hunk 12]
```diff
@@ -1,22 +1,31 @@
...
     public class ElementReferenceJsonConverterTest
     {
-        private readonly ElementReferenceJsonConverter Converter = new ElementReferenceJsonConverter();
+        private readonly IServiceProvider ServiceProvider;
+        private readonly ElementReferenceJsonConverter Converter;
+
+        public ElementReferenceJsonConverterTest()
+        {
+            ServiceProvider = Mock.Of<IServiceProvider>();
+            Converter = new ElementReferenceJsonConverter { ServiceProvider = ServiceProvider};
```
Comment (pranavkm) [inline src/Components/Server/test/ElementReferenceJsonConverterTest.cs:21]:
```suggestion
            Converter = new ElementReferenceJsonConverter { ServiceProvider = ServiceProvider };
```

File: src/Components/Web/src/ElementReferenceExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    public static class ElementReferenceExtensions
+    {
+        /// <summary>
+        /// Gives focus to an element given its <see cref="ElementReference"/>.
+        /// </summary>
+        /// <param name="elementReference">A reference to the element to focus.</param>
+        /// <returns>The <see cref="ValueTask"/> representing the asynchronous focus operation.</returns>
+        public static ValueTask FocusAsync(this ElementReference elementReference)
+        {
+            var jsRuntime = elementReference.ServiceProvider?.GetService<IJSRuntime>();
```
Comment (pranavkm) [inline src/Components/Web/src/ElementReferenceExtensions.cs:20]: It might be useful to separate out an incorrectly initialized `ElementReference` instance and a missing `IJSRuntime`. 

Something like:

```C#
if (elementReference.ServiceProvider is null)
{
   throw new ArgumentException("ElementReference has not been correctly initialized.");
}

var jsRuntime = elementReference.ServiceProvider.GetRequiredService...
```

File: src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs [hunk 14]
```diff
@@ -10,9 +11,18 @@ internal sealed class DefaultWebAssemblyJSRuntime : WebAssemblyJSRuntime
...
+
         private DefaultWebAssemblyJSRuntime()
         {
-            JsonSerializerOptions.Converters.Add(new ElementReferenceJsonConverter());
+            jsonConverter = new ElementReferenceJsonConverter();
+
+            JsonSerializerOptions.Converters.Add(jsonConverter);
+        }
+
+        internal void AddServiceProviderDependency(IServiceProvider serviceProvider)
```
Comment (pranavkm) [inline src/Components/WebAssembly/WebAssembly/src/Services/DefaultWebAssemblyJSRuntime.cs:23]: 👍

File: src/Components/test/E2ETest/Tests/ComponentRenderingTest.cs [hunk 15]
```diff
@@ -400,6 +400,26 @@ public void CanUseJsInteropToReferenceElements()
             Browser.Equal("Clicks: 2", () => inputElement.GetAttribute("value"));
         }
 
+        [Fact]
+        public void CanUseFocusExtensionToFocusElement()
```
Comment (pranavkm) [inline src/Components/test/E2ETest/Tests/ComponentRenderingTest.cs:404]: Could you add a test that calls `FocusAsync` on an `ElementReference` that is read from a JSInteop call? That should verify that the service provider is being correctly hydrated by the converter.

File: src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/ComponentsWebAssembly-CSharp.sln [hunk 16]
```diff
@@ -42,6 +30,18 @@ Global
 		{650B3CE7-2E93-4CC4-9F46-466686815EAA}.Release|x64.Build.0 = Release|Any CPU
 		{650B3CE7-2E93-4CC4-9F46-466686815EAA}.Release|x86.ActiveCfg = Release|Any CPU
 		{650B3CE7-2E93-4CC4-9F46-466686815EAA}.Release|x86.Build.0 = Release|Any CPU
+		{5990939C-7E7B-4CFA-86FF-44CA5756498A}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (pranavkm) [inline src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/ComponentsWebAssembly-CSharp.sln:33]: Could these be undone?

**Review by pranavkm — COMMENTED** [review]
Almost there!

File: src/Components/Web.JS/src/DomWrapper.ts [hunk 17]
```diff
@@ -0,0 +1,9 @@
+import '@microsoft/dotnet-js-interop';
+
+export const domFunctions = {
+  focus,
+};
+
+function focus(element: HTMLElement): void {
+  element.focus();
```
Comment (javiercn) [inline src/Components/Web.JS/src/DomWrapper.ts:8]: @SteveSandersonMS don't we have to protect here against bad elementreferences?

**Review by javiercn — COMMENTED** [review]
Overall looks great!

I think we can tweak it a bit. I don't love the part where the IServiceProvider is spilled across the application and it feels that we are missing an abstraction.

I would consider creating an ElementReferenceFactory with an implementation on M.A.C.Web that plugs in the JS runtime. The JsonConverter and the Renderer only receive a ElementReferenceFactory and the ElementReference only holds on to whatever it needs on a given platform.

We would do something like
```csharp
Renderer.ElementFactory.CreateWithUniqueId()
```
and the remote RemoteJSRuntime would simply instantiate its own ElementReferenceFactory passing itself in the constructor.

**Comment by SteveSandersonMS** [comment]
> I would consider creating an ElementReferenceFactory with an implementation on M.A.C.Web that plugs in the JS runtime

That's a bit awkward because `ElementReference` is a `struct` (and needs to be), so there's nowhere to even hold additional information. We could add some kind of `object`-valued `ExtraInfo` field onto it, but that's pretty weird.

Additionally, although for high-level features it's generally fine to push code through factories and more abstractions, the perf costs are prohibitive for code in the deep guts of rendering. There are very legit cases for building many thousands of `ElementReference`s per second (e.g., wanting to be able to reference each `td` in a grid).

I also don't really have any particular objection to referencing the `IServiceProvider` from `ElementReference`.

File: src/Components/Web.JS/src/DomWrapper.ts (hunk 17 above)
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/DomWrapper.ts:8]: 👍  Good point! It would be good to check that `element instanceof Element` or similar, throwing a clear error if not.

It will be `null` (I think) if the element it referred to has been removed from the document.

**Comment by javiercn** [comment]
> That's a bit awkward because `ElementReference` is a `struct` (and needs to be), so there's nowhere to even hold additional information. We could add some kind of `object`-valued `ExtraInfo` field onto it, but that's pretty weird and doesn't compose well if we end up needing to get access to more stuff from an `ElementReference` in the future.

It's already holding on to the IServiceProvider, so I don't see the difference here. I'm saying, initialize it with the JSRuntime only and through a Factory instead of giving it the IServiceProvider. I think having the IServiceProvider there encourages customers to do `ElementReference.ServiceProvider.GetRequiredService<PartyOn>()` which can easily get out of hand.

We designed element references to be an opaque abstraction, and we should keep treating them like that. For that reason I would have the factory live on M.A.C.Web, make it responsible for populating the state, and using an internal type that only M.A.C.Web can understand (like an internal class WebElementState) that contains the JSRuntime.

That way, the ElementReference is still opaque and we are still able to implement focus without any details leaking out.

> Additionally, although for high-level features it's generally fine to push code through factories and more abstractions, the perf costs are prohibitive for code in the deep guts of rendering. There are very legit cases for building many thousands of `ElementReference`s per second (e.g., wanting to be able to reference each `td` in a grid).

I don't think what I'm proposing would have a negative impact on perf, it would be the opposite if anything. The factory is accessed from the renderer or is part of the converter and gives clear and scoped functionality. 

Then on the ElementReference we just have state as an internal property (object) that is populated with the IJSRuntime. The helpers then just do a `if(state is IJSRuntime jsRuntime){ jsRuntime.InvokeAsync("focus",this)) which is cheaper than asking the container to resolve the jsruntime.`

In conclusion:
* The factory makes the intent clear and the functionality scoped.
* Doesn't incurs into any additional performance cost, if anything makes it faster on repeated uses.
* Prevents against pit of failure using the service locator pattern with the given service provider.

File: src/Components/Components/src/ElementReference.cs [hunk 18]
```diff
@@ -23,13 +23,24 @@ namespace Microsoft.AspNetCore.Components
...
+        /// Gets the <see cref="IServiceProvider"/> instance.
+        /// </summary>
+        public IServiceProvider ServiceProvider { get; }
+
+        /// <summary>
+        /// Instantiates a new <see cref="ElementReference" />.
+        /// </summary>
+        /// <param name="id">A unique identifier for this <see cref="ElementReference"/>.</param>
+        /// <param name="serviceProvider">The <see cref="ElementReference"/>'s <see cref="IServiceProvider"/>.</param>
+        public ElementReference(string id, IServiceProvider serviceProvider)
```
Comment (mkArtakMSFT) [inline src/Components/Components/src/ElementReference.cs:36]: @dotnet/aspnet-blazor-eng isn't this a breaking change? Is there a scenario where customers may have been using the old constructor in their code?
While necessary, I'm wondering whether we should also post an announcement about this.

Comment (pranavkm) [reply src/Components/Components/src/ElementReference.cs:36]: I guess we could keep the old ctor to avoid the breaking change. It's not particularly common to new up these instances though.

Comment (javiercn) [reply src/Components/Components/src/ElementReference.cs:36]: It's a major version, and I customers don't create this type themselves, so it's fine to break it. The major issues are unit tests that people might have written.

Comment (SteveSandersonMS) [reply src/Components/Components/src/ElementReference.cs:36]: Retaining the old constructor with `[Obsolete]` would be OK as it continue to work fine in unit tests that don't need to read back the service provider (or whatever ends up getting stored there).

However if an `[Obsolete]` also requires us to post an announcement, maybe it's easier just to do a regular breaking change.

**Comment by pranavkm** [comment]
Thanks @javiercn. Having a factory with a default that just news up the `ElementReference` seems fine. It would also clean up
`ElementReference.CreateWithUniqueId` inverting the responsibility to the caller.

a) It would be convenient if there was a default factory that constructed the type without state.

```C#
interface IElementReferenceFactory
{
   ElementReference Create();
}

ElementReference
{
  public ElementReference(string id);
  public ElementReference(string id, object? state);
}

Renderer.ElementReferenceFactory { get; } = ServiceProvider.GetService<ElementReferenceFactory> ?? DefaultElementReferenceFactory.Instance;

var newElementReference = Renderer.ElementReferenceFactory.Create();
```

It might also be useful to also add an extension to `Components.Web` that does the cast for you:

```C#
public static IJSRuntime GetJSRuntime(this ElementReference reference)
{
   return reference.State as IJSRuntime ?? throw "This hasn't been correctly configured.";
}
```

**Comment by SteveSandersonMS** [comment]
@pranavkm Let discuss all this at the sync later today! Obviously with @MackinnonBuck too.

File: src/Components/Components/src/ElementReference.cs [hunk 19]
```diff
@@ -23,13 +23,32 @@ namespace Microsoft.AspNetCore.Components
         /// </remarks>
         public string Id { get; }
 
-        public ElementReference(string id)
+        /// <summary>
+        /// Gets the nullable <see cref="ElementReferenceContext"/> instance.
```
Comment (pranavkm) [inline src/Components/Components/src/ElementReference.cs:27]:
```suggestion
        /// Gets the <see cref="ElementReferenceContext"/> instance.
```

File: src/Components/Components/src/ElementReferenceContext.cs [hunk 20]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Components
+{
+    public abstract class ElementReferenceContext
```
Comment (pranavkm) [inline src/Components/Components/src/ElementReferenceContext.cs:6]: Could you add a doc comment for this since it's public?

> Context for an `<see cref="ElementReference" />`

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 21]
```diff
@@ -76,6 +76,12 @@ public Renderer(IServiceProvider serviceProvider, ILoggerFactory loggerFactory)
         /// </summary>
         public abstract Dispatcher Dispatcher { get; }
 
+        /// <summary>
+        /// Gets or sets the <see cref="Components.ElementReferenceContext"/> associated with this <see cref="Renderer"/>,
+        /// if it exists.
+        /// </summary>
+        public ElementReferenceContext? ElementReferenceContext { get; protected set; }
```
Comment (pranavkm) [inline src/Components/Components/src/RenderTree/Renderer.cs:83]: I'm ok with this because the renderer is a fairly low level type, but you *might* be able to make this `protected internal` to avoid exposing it. It's not really useful to consumers.

File: src/Components/test/E2ETest/Tests/ComponentRenderingTest.cs (hunk 15 above)
Comment (pranavkm) [reply src/Components/test/E2ETest/Tests/ComponentRenderingTest.cs:404]: We realized this isn't a real scenario.

**Review by pranavkm — APPROVED** [review]
I'll bring this up during API review on Monday and address any feedback separately. Let's get this in to avoid further merge conflicts
